Add filter for `network ls` to hide predefined network

WeiZhang555 opened the pull request:
Add option `--no-predefined` for `network ls` to hide predefined network. 
If will be extremely useful when user want to run "docker network rm `docker network ls -q
--no-predefined`" to delete a bundle of networks, without this change, this deletion will always return non-zero exit code which usually indicates error happens because predefined network cant be removed.

Signed-off-by: Zhang Wei zhangwei555@huawei.com

Docs are still on working, will be posted after some time.

thaJeztah commented:
I understand the use case, but to be in line with other "lists", I'd prefer this to be implemented as a filter, I.e. `--filter something=...`

haven't thought about what to call the filter :-)

icecrime commented:
I understand the use case, but flags, flags, more flags, ... I don't think it's worth it =/ Sorry I'm leaning toward -1.

vdemeester commented:
I do agree with @icecrime on adding another flag, -1 on that — but I understand the use case and I like the idea of the `--filter` flags as it's already present in other *listing* commands (`images`, `ps`) so it does feel more *consistent* (and extensible :stuck_out_tongue_closed_eyes:).

The name of the filter could be `type` and values would be `built-in`, etc..

icecrime commented:
@WeiZhang555 Would you be willing to update to use a `--filter=...` syntax? Thanks!

WeiZhang555 commented:
Yeah, I can update it with a `--filter=...`, but before doing that, I want to say something more for myself.

Actually I have considered these two very different implementations, I mean *one more option* vs *`--filter`*, but as `docker network ls` gives only three fields: `container id`, `name` and `driver`, it show very limited information for us to manipulate, `--filter` is powerful of course, but I don't think it needs so powerful a option.

What's more is that comparing to `docker image`, we have *docker rmi `docker images -aq`*, you see that it's quite short and easy to type, I prefer to have one option to let me run *docker network rm `docker network ls -u`*  (`-u` can stand for *--user-defined*, my original `--no-predefined` is too long and I'm not satisfied to it either.).  The key point is that I think it's worth a short flat format such as (*docker network rm `docker network ls -u`*) instead of a long format (*docker network rm `docker network ls --filter=built-in`*).

As for consistency, (*docker rm `docker ps -aq`*) is also a short command similar to (*docker rmi `docker images -aq`*)

However, I won't argue if you all think `--filter` is better, so are you still insist on option of `--filter` ? :smile: 
Whatever thank you for your kind suggestions! @thaJeztah @icecrime @vdemeester

WeiZhang555 commented:
I just renamed original option `--no-predefined` to `-u/--user-defined`, so user can run `docker network rm 'docker network ls -qu'` to delete a bundle of networks. 
This is more friendly than something like `docker network rm 'docker network ls -q --filter type=custom'` for lazy person like me :stuck_out_tongue: 

But I'm still waiting for your opinions, and I won't mind to change is to filter format if you insist.  :smile:

WeiZhang555 committed:
Add filter for `network ls` to hide predefined net

Add filter support for `network ls` to hide predefined network,
then user can use "docker network rm `docker network ls -f type=custom`"
to delete a bundle of userdefined networks.

Signed-off-by: Zhang Wei <zhangwei555@huawei.com>

WeiZhang555 commented:
I just re-implemented `-u` option with `--filter` for hiding predefined network,
now user can use `docker network rm 'docker network ls -f type=custom'`
to delete a bundle of userdefined networks.

Currently filter support only two options:
type=custom : only show userdefined networks
type=built-in: only show built in networks

Final behavior look likes this:
```
$ docker network ls       (# normal network ls)
NETWORK ID          NAME                DRIVER
e94695c3f8a2        none                null
3841be13515b        host                host
e397be974220        dev                 bridge
aad8b004629c        dev1                bridge
f71a378fc0c8        bridge              bridge

$ docker network ls -f type=custom       (# only show user defined network)
NETWORK ID          NAME                DRIVER
e397be974220        dev                 bridge
aad8b004629c        dev1                bridge

$ docker network ls -f type=built-in       (# only show built-in network)
NETWORK ID          NAME                DRIVER
3841be13515b        host                host
f71a378fc0c8        bridge              bridge
e94695c3f8a2        none                null

$ docker network ls -f type=built-in -f type=custom (# compose two kinds of filter equals to normal ls)
NETWORK ID          NAME                DRIVER
f71a378fc0c8        bridge              bridge
e94695c3f8a2        none                null
3841be13515b        host                host
e397be974220        dev                 bridge
aad8b004629c        dev1                bridge

```

thaJeztah commented:
Thanks @WeiZhang555! I understand that "shorter" options can be convenient, but I think we should be consistent; also `--filter` is more flexible for future extension (e.g. `--filter driver=overlay`). So thanks for implementing as a `--filter` :+1: 

design (UX) LGTM

WeiZhang555 commented:
Docs updated.
I think It would be better to merge #17489 first before this, because in docs I said 
" `--filter type=custom` is extremely useful when you are trying to delete a bundle of networks"
This is true only when `docker network rm` can remove many networks at one time, that's why #17489 must be done first. :smile:

vdemeester commented:
@WeiZhang555 The UX looks good to me but this should be implemented on the "server" side, as anything the `cli` does should be do-able using the API. Thus, the code in `api/client/network.go` should be really small (and more or less look like the filter part of `api/client/ps.go`) and the validation and actual filtering should be done server side.

Once this is done, I'll put this into code review :wink:.

WeiZhang555 commented:
@vdemeester OK, I'll move the filter function to server side, maybe after I come back from DockerCon in Europe. :stuck_out_tongue_winking_eye:

thaJeztah commented:
Nice! In the airplane now to Barcelona, so see you there! (@vdemeester will be there as well \o/)

WeiZhang555 commented:
@thaJeztah @vdemeester Looking forward to seeing you guys :laughing:

calavera commented on api/client/network.go:
```diff
@@ -14,9 +14,26 @@ import (
+
+var supportedFilters map[string]filterHandlers
+
+func init() {
```
you can initialize those supported filters directly when you create the variable, two lines above. This init function is not necessary.

calavera commented:
I might be wrong, but I feel like the filter validations should be on the server side. Otherwise, people using the API won't have them.

thaJeztah commented:
agreed with @calavera, validations should be server side

WeiZhang555 commented:
Hello everyone, I have reimplemented this function on server-side, I think it should work fine .
Now there are two more filters:
```
-f type=builtin
-f type-=custom
```
Hmm, there are already two existing filters though not enabled, I'm not sure who are using it. They are 
```
-f id=XXX
-f name=XXX
```
I don't know what it's designed for, so I just keep it there.
Tell me what do you think of this, thank you all! :smile:

calavera commented on api/server/router/network/network_routes.go:
```diff
@@ -17,6 +17,68 @@ import (
 )
 
+// supportedFilters predefined some supported filter handler function
+type filterHandler func(libnetwork.Network, []string) (bool, error)
```
can we put all this filtering logic in a separated file, like `api/server/router/network/filters.go`? just to keep them separated from the routes :smile:

calavera commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -257,6 +257,31 @@ func (s *DockerNetworkSuite) TestDockerNetworkLsDefault(c *check.C) {
 	}
 }
 
+func (s *DockerNetworkSuite) TestDockerNetworkLsWithNoPredefined(c *check.C) {
```
we should add a test for the type `builtin`, checking that custom networks don't show up.

calavera commented:
this is looking great :+1: just a couple of small improvements.

WeiZhang555 commented:
@calavera Nice suggestion, I have made relative modification following your comment, please take a look at it. Thank you! :)

cpuguy83 commented on integration-cli/docker_cli_network_unix_test.go:
```diff
@@ -257,6 +257,93 @@ func (s *DockerNetworkSuite) TestDockerNetworkLsDefault(c *check.C) {
+	var hostExists, noneExists, bridgeExists, devExists bool
+	for _, line := range lines[1 : len(lines)-1] {
+		netName := strings.Fields(line)[1]
+		if netName == "dev" {
```
can we use a switch here?

```go
switch netMode := runconfig.NetworkMode(netName) {
case neMode.IsBridge():
    bridgeExists = true
case netMode.IsHost():
    hostExists = true
case netMode.IsNone():
    noneExists = true
case netMode.IsUserDefined():
    if netName == "dev" {
        devExists = true
    }
}
```

cpuguy83 replied:
Also, this is duplicated from above as well, might be good to extract into a helper.

WeiZhang555 commented:
@cpuguy83 I re-implemented the test  case and try a new method to squash 3 test cases into one with helper function. I think it looks better now. PTAL. :smile:

WeiZhang555 commented:
ping @cpuguy83 ...

cpuguy83 commented:
LGTM pending tests

calavera commented:
LGTM. Moving to docs review.

WeiZhang555 commented:
ping @thaJeztah ...

vdemeester commented:
@WeiZhang555 needs a rebase :sweat:

thaJeztah commented:
Oh, @WeiZhang555 sorry for the delay, I'll review this first thing tomorrow :blush:

WeiZhang555 commented:
@calavera this PR conflicts with your previous PR #18266 while I'm doing the rebasing, then I found that  you refactor the filter package and make a great performance improvement, so I refactored my code to adapt your new improvement(mainly introduced `netFilters.WalkValues` and `netFilters.Validate`). 

Maybe we need to return back to `2-code-review` because the code has changed, could you help me check whether I make it right or not according to your design for filter? @calavera 

Also ping other guys, @cpuguy83 @vdemeester , sorry for bothering...

vdemeester commented:
@WeiZhang555 good point, putting it back into code-review.
LGTM for me

thaJeztah commented:
> sorry for bothering...

No, please don't say "sorry", that's excellent, thanks for catching that and updating your code, awesome! :+1:

cpuguy83 commented on api/server/router/network/network_routes.go:
```diff
@@ -28,29 +27,24 @@ func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWrit
-		for _, nw := range nwList {
-			list = append(list, buildNetworkResource(nw))
-		}
+	nwList := n.backend.GetNetworksByID("")
```
I'm a little suspect of this function.
I'd prefer to have a `GetAllNetworks` rather than passing an empty string to `GetNetworksByID`

cpuguy83 commented on api/server/router/network/network_routes.go:
```diff
@@ -28,29 +27,24 @@ func (n *networkRouter) getNetworksList(ctx context.Context, w http.ResponseWrit
-			list = append(list, buildNetworkResource(nw))
-		}
+	nwList := n.backend.GetNetworksByID("")
+	displayable, err := filterAllNetworks(nwList, netFilters)
```
And then call this `filterNetworks`

cpuguy83 commented:
Couple of minor nits.

WeiZhang555 commented:
@cpuguy83 code updated as per your suggestion, please take a look, thank you :smile:

cpuguy83 commented:
LGTM, moving to docs

tiborvass commented:
@WeiZhang555 needs a rebase sorry :(

WeiZhang555 commented:
@tiborvass Never mind. :smile: 
Rebased.

thaJeztah commented:
Working on some suggestions for the docs, will post them later today :+1:

thaJeztah commented on api/server/router/network/filter.go:
```diff
@@ -0,0 +1,99 @@
+
+func filterNetworkByName(nws []libnetwork.Network, name string) (retNws []libnetwork.Network, err error) {
+	for _, nw := range nws {
+		if nw.Name() == name {
```
@WeiZhang555 while looking at the documentation, I noticed that `--filter name=..` only allows filtering on a *full* match, correct? Because we do allow filtering on partial matches for, e.g., containers; https://github.com/docker/docker/blob/33ab2bb52c130380e038013d68fdd8ad3c663360/pkg/parsers/filters/parse.go#L174

wondering if we want this filter to work the same?

WeiZhang555 replied:
@thaJeztah Currently yes, it only allows filtering on a full match of name.

I'm not sure whether we need a partial match behavior, actually I have no idea why there is a name filter here and what it's used for, because it's never enabled from Docker's client, and only reachable via direct API visit. So I simply keep it exactly the same as before, adding one more `type` filter.

Do you think we need a partial match?

thaJeztah replied:
@WeiZhang555 eventually I think it would be good to be consistent, and also do partial matches. I think we should do that in a follow up. :smile:

WeiZhang555 replied:
@thaJeztah  OK. I would fix that in another PR. Let's focus on the `type` filter for now.
WDYT?

calavera commented:
This implementation needs to change quite a lot after merging the new client library. I'm happy to assist or carry the changes if needed. If we can finalize the documentation the process can be slightly faster.

@WeiZhang555 don't hesitate to ping me if you need help.

WeiZhang555 commented:
@calavera  Thank you! 
I'll ping you as soon as the docs are all well arranged and rebasing starts.

thaJeztah commented on docs/reference/commandline/network_ls.md:
````diff
@@ -38,8 +39,40 @@ NETWORK ID                                                         NAME
+63d1ff1f77b07ca51070a8c227e962238358bd310bde1529cf62e6c307ade161   dev                 bridge
 ```
 
+`--filter` is used to get rid of items user doesn't care from results, for
````
I used the "filtering" section of the `docker ps` documentation to make it more consistent, wdyt?

    ## Filtering

    The filtering flag (`-f` or `--filter`) format is a `key=value` pair. If there
    is more than one filter, then pass multiple flags (e.g. `--filter "foo=bar" --filter "bif=baz"`).
    Multiple filter flags are combined as an `OR` filter. For example, 
    `-f type=custom -f type=builtin` returns both `custom` and `builtin` networks.

    The currently supported filters are:

    * id (network's id)
    * name (network's name)
    * type (custom|builtin)


    #### Type

    The `type` filter supports two values; `builtin` displays predefined networks
    (`bridge`, `none`, `host`), whereas `custom` displays user defined networks.

    The following filter matches all user defined networks:

    ```bash
    $ docker network ls --filter type=custom
    NETWORK ID          NAME                DRIVER
    95e74588f40d        foo                 bridge
    63d1ff1f77b0        dev                 bridge
    ```

    By having this flag it allows for batch cleanup. For example, use this filter
    to delete all user defined networks:

    ```bash
    $ docker network rm `docker network ls --filter type=custom -q`
    ```

    A warning will be issued when trying to remove a network that has containers
    attached.

    #### Name

    The `name` filter matches on all or part of a network's name.

    The following filter matches all networks with a name matching the `foo` string.

    ```bash
    $ docker network ls --filter name=foo
    NETWORK ID          NAME                DRIVER
    95e74588f40d        foo                 bridge
    ```

    #### ID

    The `id` filter matches on all or part of a network's ID.

    The following filter matches all networks with a name containing the
    `06e7eef01700` string.

    ```bash
    $ docker network ls --filter id=63d1ff1f77b07ca51070a8c227e962238358bd310bde1529cf62e6c307ade161
    NETWORK ID          NAME                DRIVER
    63d1ff1f77b0        dev                 bridge
    ```

    You can also filter for a substring in a ID as this shows:

    ```bash
    $ docker ps --filter id=95e74588f40d
    NETWORK ID          NAME                DRIVER
    95e74588f40d        foo                 bridge

    $ docker ps --filter id=95e
    NETWORK ID          NAME                DRIVER
    95e74588f40d        foo                 bridge
    ```

WeiZhang555 replied:
This is great! :+1:

WeiZhang555 commented:
@thaJeztah docs updated. PTAL.

cpuguy83 commented:
Looks like a bad rebase

WeiZhang555 commented:
@cpuguy83 Yeah, you're right. I only updated the docs, and I will do the code rebase ASAP. :smile:

WeiZhang555 commented:
Rebased on the new client library @calavera 

And @thaJeztah , partial name match is implemented on https://github.com/docker/docker/pull/17782/files#diff-c5b0cdb16e32c93e45e51c5131209cbaR68, with relevant test case change on https://github.com/docker/docker/pull/17782/files#diff-401699de8561309143baebae1c3696e3R289. Also docs are updated as per your suggestions.

Also ping other guys involved, please take a look. :smile:

vdemeester commented:
Still LGTM for me :wink:

thaJeztah commented on docs/reference/commandline/network_ls.md:
````diff
@@ -38,8 +39,82 @@ NETWORK ID                                                         NAME
+$ docker network ls --filter name=foo
+NETWORK ID          NAME                DRIVER
+95e74588f40d        foo                 bridge
+```
````
Now that we have partial matches, can you update this to:

    #### Name

    The `name` filter matches on all or part of a network's name.

    The following filter matches all networks with a name containing the `foobar` string.

    ```bash
    $ docker network ls --filter name=foobar
    NETWORK ID          NAME                DRIVER
    06e7eef0a170        foobar              bridge
    ```

    You can also filter for a substring in a name as this shows:

    ```bash
    $ docker ps --filter name=foo
    NETWORK ID          NAME                DRIVER
    95e74588f40d        foo                 bridge
    06e7eef0a170        foobar              bridge
    ```

thaJeztah commented:
@WeiZhang555 just realized; this is now done server-side? If so, we need to update the API docs as well;

[In the API changelog](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v122-api-changes), add something like:

    `GET /networks` now supports filtering by `name`, `id` and `type`.

The  [`GET /networks` endpoint in the API docs](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.22.md#list-networks) also need a mention of the filter. [Similar to "list containers"](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.22.md#list-containers)

WeiZhang555 commented:
@thaJeztah docs updated.

And I happen to find that all the filter description is outdated, e.g. 
```
**filters** - JSON encoded value of the filters (a `map[string][]string`) to process 
```

It's not a `map[string][]string` any more, it's a `map[string]map[string]bool` for now, the change is introduced by #18266. That's why I updated all filter parts, not many changes, please take a look. :smile:

thaJeztah commented:
> It's not a map[string][]string any more, it's a map[string]map[string]bool for now, the change is introduced by #18266. That's why I updated all filter parts, not many changes, please take a look. :smile:

ping @calavera can you check that, is that a deliberate change, because that isn't documented (and sounds like a breaking change in the API)

WeiZhang555 commented:
ping @calavera

calavera commented:
> sounds like a breaking change in the API

the api still supports `map[string][]string`. We don't support it internally anymore. I'm okay with leaving the old description if it's more clear and causes less confusion.

https://github.com/calavera/docker/blob/93d1dd8036d57f5cf1e5cbbbad875ae9a6fa6180/pkg/parsers/filters/parse.go#L84-L91

thaJeztah commented:
> I'm okay with leaving the old description if it's more clear and causes less confusion.

hm, perhaps it's better to keep the old description then. not sure we want to describe that "you can either formatA or formatB"

Would you be okay changing those descriptions back @WeiZhang555?

WeiZhang555 commented:
@thaJeztah Filter descriptions have been changed back already, please take a look. :)

thaJeztah commented:
Thanks @WeiZhang555!

LGTM

ping @moxiegirl for a final look :+1:

moxiegirl commented on docs/reference/api/docker_remote_api_v1.22.md:
```diff
@@ -2794,11 +2794,12 @@ Content-Type: application/json
 Query Parameters:
 
-- **filters** - JSON encoded value of the filters (a `map[string][]string`) to process on the networks list. Available filters: `name=[network-names]` , `id=[network-ids]`
+- **filters** - JSON encoded value of the filters (a `map[string][]string`) to process on the networks list. Available filters:
```
@WeiZhang555 I'm not sure it is necessary to mention the format if there are only two types of filters available. How about:

```markdown
- **filters** - JSON encoded network list filter. The value is one of: `name=[network-names]` or `id=[network-ids]`
```

thaJeztah replied:
It's consistent with the way it's documented for other filters though, e.g. For [list containers](https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api_v1.22.md#list-containers). Perhaps we should rewrite them all, and provide some actual examples

WeiZhang555 commented:
@moxiegirl docs updated.

thaJeztah commented:
Thanks @WeiZhang555, given that those changes are what @moxiegirl suggested, I take that as an implicit "LGTM" from @moxiegirl, and I think this is ready to merge!

LGTM from me as well

thanks for your contribution, @WeiZhang555!!
